File: src/test/java/com/github/alexandrecarlton/idea/settings/applier/impl/configurations/docker/DockerComposeConfigurationSettingsApplierTest.java
package com.github.alexandrecarlton.idea.settings.applier.impl.configurations.docker;

import static org.assertj.core.api.Assertions.assertThat;

import com.github.alexandrecarlton.idea.settings.applier.api.SettingsApplier;
import com.github.alexandrecarlton.idea.settings.fixtures.IdeaSettingsTestFixture;
import com.github.alexandrecarlton.idea.settings.layout.configurations.docker.DockerComposeConfigurationSettings;
import com.github.alexandrecarlton.idea.settings.layout.configurations.docker.ImmutableDockerComposeConfigurationOptionsSettings;
import com.github.alexandrecarlton.idea.settings.layout.configurations.docker.ImmutableDockerComposeConfigurationSettings;
import com.github.alexandrecarlton.idea.settings.layout.configurations.docker.ImmutableDockerEnvironmentVariable;
import com.google.common.collect.ImmutableList;
import com.intellij.docker.DockerDeploymentConfiguration;
import com.intellij.docker.DockerRunConfigurationCreator;
import com.intellij.docker.agent.settings.DockerEnvVarImpl;
import com.intellij.execution.RunManager;
import com.intellij.execution.RunnerAndConfigurationSettings;
import com.intellij.remoteServer.impl.configuration.deployment.DeployToServerRunConfiguration;
import java.nio.file.Paths;
import java.util.Collections;
import org.junit.Before;
import org.junit.Test;

public class DockerComposeConfigurationSettingsApplierTest extends IdeaSettingsTestFixture {

  private DockerRunConfigurationCreator dockerRunConfigurationCreator;
  private SettingsApplier<DockerComposeConfigurationSettings> settingsApplier;
  private RunManager runManager;


  @Override
  @Before
  public void setUp() throws Exception {
    super.setUp();
    dockerRunConfigurationCreator = new DockerRunConfigurationCreator(project);
    runManager = RunManager.getInstance(project);
    settingsApplier = new DockerComposeConfigurationSettingsApplier(dockerRunConfigurationCreator, runManager);
  }

  @Test
  public void servicesApplied() {
    settingsApplier.apply(ImmutableDockerComposeConfigurationSettings.builder()
        .name("Docker Compose Services")
        .services(Collections.singletonList("service1"))
        .build());
    assertThat(getDockerDeploymentConfiguration("Docker Compose Services").getServices())
        .containsExactly("service1");
  }

  @Test
  public void composeFilesApplied() {
    settingsApplier.apply(ImmutableDockerComposeConfigurationSettings.builder()
        .name("Docker Compose Files")
        .composeFiles(ImmutableList.of(
            Paths.get(project.getBasePath()).resolve("docker-compose.yml"),
            Paths.get(project.getBasePath()).resolve("docker-compose-2.yml"),
            Paths.get(project.getBasePath()).resolve("docker-compose-3.yml")))
        .build());
    assertThat(getDockerDeploymentConfiguration("Docker Compose Files").getSourceFilePath())
        .isEqualTo(Paths.get(project.getBasePath()).resolve("docker-compose.yml").toString());
    assertThat(getDockerDeploymentConfiguration("Docker Compose Files").getSecondarySourceFiles())
        .containsExactly(
            Paths.get(project.getBasePath()).resolve("docker-compose-2.yml").toString(),
            Paths.get(project.getBasePath()).resolve("docker-compose-3.yml").toString());
  }

  @Test
  public void environmentVariablesApplied() {
    settingsApplier.apply(ImmutableDockerComposeConfigurationSettings.builder()
        .name("Docker Compose Environment Variables")
        .environmentVariables(ImmutableList.of(
            ImmutableDockerEnvironmentVariable.builder()
                .name("name")
                .value("value")
                .build()))
        .build());
    assertThat(getDockerDeploymentConfiguration("Docker Compose Environment Variables").getEnvVars())
        .containsExactly(new DockerEnvVarImpl("name", "value"));
  }

  @Test
  public void forceBuildApplied() {
    settingsApplier.apply(ImmutableDockerComposeConfigurationSettings.builder()
        .name("Docker Compose Force Build")
        .options(ImmutableDockerComposeConfigurationOptionsSettings.builder()
            .buildForceBuildImages(true)
            .build())
        .build());
    assertThat(getDockerDeploymentConfiguration("Docker Compose Force Build").getRunCliOptions())
        .isEqualTo("--build");
  }

  private DockerDeploymentConfiguration getDockerDeploymentConfiguration(String name) {
    RunnerAndConfigurationSettings runnerAndConfigurationSettings = runManager.findConfigurationByName(name);
    assertThat(runnerAndConfigurationSettings).isNotNull();
    DeployToServerRunConfiguration deployToServerRunConfiguration = (DeployToServerRunConfiguration) runnerAndConfigurationSettings.getConfiguration();
    return (DockerDeploymentConfiguration) deployToServerRunConfiguration.getDeploymentConfiguration();
  }
}

File: src/test/kotlin/com/github/alexandrecarlton/idea/settings/applier/impl/configurations/docker/DockerComposeConfigurationSettingsApplierTest.kt
package com.github.alexandrecarlton.idea.settings.applier.impl.configurations.docker

import com.github.alexandrecarlton.idea.settings.applier.api.SettingsApplier
import com.github.alexandrecarlton.idea.settings.fixtures.IdeaSettingsTestFixture
import com.github.alexandrecarlton.idea.settings.layout.configurations.DockerComposeConfigurationSettings
import com.github.alexandrecarlton.idea.settings.layout.configurations.docker.DockerComposeConfigurationOptionsSettings
import com.github.alexandrecarlton.idea.settings.layout.configurations.docker.DockerEnvironmentVariable
import com.intellij.docker.DockerDeploymentConfiguration
import com.intellij.docker.DockerRunConfigurationCreator
import com.intellij.docker.agent.settings.DockerEnvVarImpl
import com.intellij.execution.RunManager
import com.intellij.remoteServer.impl.configuration.deployment.DeployToServerRunConfiguration
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import java.io.File
import java.nio.file.Paths

class DockerComposeConfigurationSettingsApplierTest : IdeaSettingsTestFixture() {

    private lateinit var dockerRunConfigurationCreator: DockerRunConfigurationCreator
    private lateinit var settingsApplier: SettingsApplier<DockerComposeConfigurationSettings>
    private lateinit var runManager: RunManager

    @Before
    public override fun setUp() {
        dockerRunConfigurationCreator = DockerRunConfigurationCreator(project)
        runManager = RunManager.getInstance(project)
        settingsApplier = DockerComposeConfigurationSettingsApplier(dockerRunConfigurationCreator, runManager)
    }

    @Test
    fun servicesApplied() {
        settingsApplier.apply(DockerComposeConfigurationSettings(
            name = "Docker Compose Services",
            services = listOf("service1")))
        assertThat(getDockerDeploymentConfiguration("Docker Compose Services").services).containsExactly("service1")
    }

    @Test
    fun composeFilesApplied() {
        settingsApplier.apply(DockerComposeConfigurationSettings(
            name = "Docker Compose Files",
            composeFiles = listOf(
                File(project.basePath).resolve("docker-compose.yml"),
                File(project.basePath).resolve("docker-compose-2.yml"),
                File(project.basePath).resolve("docker-compose-3.yml"))))
        assertThat(getDockerDeploymentConfiguration("Docker Compose Files").sourceFilePath)
                .isEqualTo(Paths.get(project.basePath).resolve("docker-compose.yml").toString())
        assertThat(getDockerDeploymentConfiguration("Docker Compose Files").secondarySourceFiles)
                .containsExactly(
                        Paths.get(project.basePath).resolve("docker-compose-2.yml").toString(),
                        Paths.get(project.basePath).resolve("docker-compose-3.yml").toString())
    }

    @Test
    fun environmentVariablesApplied() {
        settingsApplier.apply(DockerComposeConfigurationSettings(
            name = "Docker Compose Environment Variables",
            environmentVariables = listOf(DockerEnvironmentVariable(name = "name", value = "value"))))
        assertThat(getDockerDeploymentConfiguration("Docker Compose Environment Variables").envVars)
                .containsExactly(DockerEnvVarImpl("name", "value"))
    }

    @Test
    fun forceBuildApplied() {
        settingsApplier.apply(DockerComposeConfigurationSettings(
            name = "Docker Compose Force Build",
            options = DockerComposeConfigurationOptionsSettings(
                buildForceBuildImages = true)))
        assertThat(getDockerDeploymentConfiguration("Docker Compose Force Build").runCliOptions)
                .isEqualTo("--build")
    }

    private fun getDockerDeploymentConfiguration(name: String): DockerDeploymentConfiguration {
        val runnerAndConfigurationSettings = runManager.findConfigurationByName(name)
        assertThat(runnerAndConfigurationSettings).isNotNull()
        val deployToServerRunConfiguration = runnerAndConfigurationSettings!!.configuration as DeployToServerRunConfiguration<*, *>
        return deployToServerRunConfiguration.deploymentConfiguration as DockerDeploymentConfiguration
    }
}
